### ISSUE TITLE
Adding Auditlog to Installed Apps causes PostreSQL JSON fields to be re-encoded on model save, without explicity registering the model

### ISSUE THREAD

**Issue by Jackevansevo** [jjkester/django-auditlog#71]
Following Django's documentation for PostgreSQL's JSON field: https://docs.djangoproject.com/en/1.10/ref/contrib/postgres/fields/#querying-jsonfield

I get some weird behaviour after including AuditLog in my installed apps:

**#models.py**
```python
from django.contrib.postgres.fields import JSONField
from django.db import models


class Dog(models.Model):
    name = models.CharField(max_length=200)
    data = JSONField()

    def __str__(self):  # __unicode__ on Python 2
        return self.name
```

**#settings.py**
```python
DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.postgresql_psycopg2',
        'NAME': 'audittest',
        'USER': 'audittest',
        'PASSWORD': 'audittest',
        'HOST': 'localhost',
        'PORT': '',
    }
}
```

**# pip freeze** 
```bash
(venv) jwe in ~/Documents/audittest λ pip freeze 
Django==1.10.1
django-auditlog==0.4.0
django-jsonfield==1.0.1
psycopg2==2.6.2
```

**./manage.py shell**
```python
(venv) jwe in ~/Documents/audittest λ ./manage.py shell
Python 3.4.2 (default, Oct  8 2014, 10:45:20) 
[GCC 4.9.1] on linux
Type "help", "copyright", "credits" or "license" for more information.
(InteractiveConsole)
>>> from jsontest.models import Dog
>>> Dog.objects.create(name='Meg', data={'breed': 'collie'})
<Dog: Meg>
>>> Dog.objects.first().data
'{"breed": "collie"}'
>>> Dog.objects.first().save()
>>> Dog.objects.first().data
'"{\\"breed\\": \\"collie\\"}"'
>>> Dog.objects.first().save()
>>> Dog.objects.first().data
'"\\"{\\\\\\"breed\\\\\\": \\\\\\"collie\\\\\\"}\\""'
>>> Dog.objects.first().save()
>>> Dog.objects.first().data
'"\\"\\\\\\"{\\\\\\\\\\\\\\"breed\\\\\\\\\\\\\\": \\\\\\\\\\\\\\"collie\\\\\\\\\\\\\\"}\\\\\\"\\""'
>>> 
```

**Things to note:**

-  I did not include `'auditlog.middleware.AuditlogMiddleware'`
- Issue occurs both before and after I run `./manage.py migrate`
- I did not register the model to log for changes with `auditlog.register(Dog)`

I'll try snooing around the code base this weekend to figure out what's going on, but I can't garuntee anything. Just thought I make this known incase any other people encouter the issue. 

Cheers, Jack

**Comment by jjkester** [comment]
Are you sure that auditlog caused this weird behaviour? I can personally guarantee that the Auditlog code does not "hack" (or: monkey patch) any Django code. I recommend you try this again with only `django-jsonfield` installed.

**Comment by Jackevansevo** [comment]
I'm probably completely missing something here...

Starting afresh with just django-jsonfield I get the following behaviour

```bash
(venv) jwe in ~/Documents/audittest λ pip freeze
Django==1.10.1
psycopg2==2.6.2
(venv) jwe in ~/Documents/audittest λ pip install django-jsonfield 
Collecting django-jsonfield
Installing collected packages: django-jsonfield
Successfully installed django-jsonfield-1.0.1
(venv) jwe in ~/Documents/audittest λ pip freeze
Django==1.10.1
django-jsonfield==1.0.1
psycopg2==2.6.2
(venv) jwe in ~/Documents/audittest λ ./manage.py shell
Python 2.7.9 (default, Mar  1 2015, 12:57:24) 
[GCC 4.9.2] on linux2
Type "help", "copyright", "credits" or "license" for more information.
(InteractiveConsole)
>>> from jsontest.models import Dog
>>> Dog.objects.create(name='Meg', data={'breed': 'collie'})
<Dog: Meg>
>>> Dog.objects.first().data
{u'breed': u'collie'}
>>> Dog.objects.first().save()
>>> Dog.objects.first().data
{u'breed': u'collie'}
>>> 
```

This version seems to be the exact same version pulled by django-auditlog, but the behaviour I experience is different. Am I missing a step?

```bash
(venv) jwe in ~/Documents/audittest λ pip freeze
Django==1.10.1
psycopg2==2.6.2
(venv) jwe in ~/Documents/audittest λ pip install django-auditlog
Collecting django-auditlog
Requirement already satisfied (use --upgrade to upgrade): Django>=1.8 in ./venv/lib/python2.7/site-packages (from django-auditlog)
Collecting django-jsonfield>=1.0.0 (from django-auditlog)
Installing collected packages: django-jsonfield, django-auditlog
Successfully installed django-auditlog-0.4.0 django-jsonfield-1.0.1
(venv) jwe in ~/Documents/audittest λ pip freeze
Django==1.10.1
django-auditlog==0.4.0
django-jsonfield==1.0.1
psycopg2==2.6.2

# After adding 'auditlog' to installed apps and migrating

(venv) jwe in ~/Documents/audittest λ ./manage.py shell
Python 2.7.9 (default, Mar  1 2015, 12:57:24) 
[GCC 4.9.2] on linux2
Type "help", "copyright", "credits" or "license" for more information.
(InteractiveConsole)
>>> from jsontest.models import Dog
>>> Dog.objects.create(name='Meg', data={'breed': 'collie'})
<Dog: Meg>
>>> Dog.objects.first().data
'{"breed": "collie"}'
>>> Dog.objects.first().save()
>>> Dog.objects.first().data
'"{\\"breed\\": \\"collie\\"}"'
>>> 
```

**Comment by jjkester** [comment]
Maybe I'm missing something here. Auditlog should *never* change the contents of model fields, especially when a model is not registered to be logged.

Looking at the 'auditlog-included' example it seems that the data is not decoded but continuously re-encoded to json, which is very strange.

One thing I can say is that Django's postgres-specific `JSONField` class has the same name as the one from `django-jsonfield`, but that really should not be a problem considering their import paths are different.

To summarize, I have no clue what's going on here. Additionally, I'm quite busy so I would appreciate if you want to look into this.

---

As I am writing this I suddenly realize that importing something from the `django-jsonfield` package could be key here. At least, it is a factor which differs between your two examples. Maybe this package alters the default field for some reason? (That would be very bad though...)

**Comment by Jackevansevo** [comment]
I'll take a snoop around and see if I can find anything this weekend. All signs point towards some weird package name conflict behaviour

Cheers, Jack

**Comment by dev-drprasad** [comment]
I am also having same issue

**Comment by Seleznev-nvkz** [comment]
I have the same problem :(

**Comment by kbussell** [comment]
The problem is that if your project uses django-jsonfield (which django-auditlog uses), you cannot use Django's native JSONField anywhere else in your project. For more details, see https://bitbucket.org/schinckel/django-jsonfield/issues/57/cannot-use-in-the-same-project-as-djangos.

I had been meaning to figure out a decent solution to this problem, as several 3rd-party libraries we bring in use django-jsonfield for better portability across databases.

**Comment by kbussell** [comment]
I created an experimental PR #81 to fix this issue. After using this branch, add
```
USE_NATIVE_JSONFIELD = True
```
to your project's settings, and then run `python manage.py migrate`.

This will update auditlog's LogEntry model to be a native JSONField type. You can switch everything back by changing the above setting to `False`, just be sure to run migrations afterwards.

The underlying compatibility layer has not been fully tested yet, so please don't run this in production until you're sure it works for your use case!

**Comment by jjkester** [comment]
@kbussell I have copied your comments about your solution to the PR description as well.

**Comment by eriktelepovsky** [comment]
Can somebody explain, how to use native JSON fields of Postgres in custom models while django-auditlog app is installed without touching its code, please?

**Comment by huix11** [comment]
ok I know
